### PR TITLE
feat(ErrorBoundary): add ErrorBoundary around all modules

### PR DESCRIPTION
**Pull request by JamesSingleton**
This wraps modules being added to the module map with an ErrorBoundary HoC

### CODE REVIEW

File: packages/holocron/src/HolocronErrorBoundary.jsx [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+  render() {
+    const { props: { children }, state: { error } } = this;
+
+    if (error) {
+      if (process.env.NODE_ENV === 'production') return null;
+      return <div style={{ color: 'red' }}>{error.message}</div>;
+    }
+
+    return children;
```
Comment (10xLaCroixDrinker) [inline packages/holocron/src/HolocronErrorBoundary.jsx]: this won't work if there are multiple children

File: packages/holocron/src/HolocronErrorBoundary.jsx [hunk 2]
```diff
@@ -0,0 +1,58 @@
+import React, { Component } from 'react';
+import { node } from 'prop-types';
```
Comment (10xLaCroixDrinker) [inline packages/holocron/src/HolocronErrorBoundary.jsx]: a little unorthodox but ¯\_(ツ)_/¯

File: packages/holocron/__tests__/HolocronErrorBoundary.spec.jsx [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+    ).toBe(null);
+
+    expect(
+      wrapperWithHolocronErrorBoundary.contains(
+        <CharacterWelcome name="Chewbacca" />
+      )
+    ).toBe(true);
+  });
+
+  it('Sets its state t an error state and renders the error message', () => {
```
Comment (10xLaCroixDrinker) [inline packages/holocron/__tests__/HolocronErrorBoundary.spec.jsx]:
```suggestion
  it('Sets its state to an error state and renders the error message', () => {
```

File: packages/holocron/__tests__/HolocronErrorBoundary.spec.jsx [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+      <CharacterWelcomeWithHolocronErrorBoundary name="R2-D2" />
+    );
+
+    expect(wrapper.state().error).toEqual(expect.objectContaining(mockError));
+    expect(
+      wrapperWithHolocronErrorBoundary.find(HolocronErrorBoundary).instance().state.error
+    ).toEqual(expect.objectContaining(mockError));
+  });
+
+  it('Sets its state t an error state and renders null in production', () => {
```
Comment (10xLaCroixDrinker) [inline packages/holocron/__tests__/HolocronErrorBoundary.spec.jsx]:
```suggestion
  it('Sets its state to an error state and renders null in production', () => {
```

File: packages/holocron/__tests__/HolocronErrorBoundary.spec.jsx (hunk 3 above)
Comment (10xLaCroixDrinker) [reply packages/holocron/__tests__/HolocronErrorBoundary.spec.jsx]: I dont see any assertions on the render as the spec suggests

File: packages/holocron/__tests__/HolocronErrorBoundary.spec.jsx (hunk 4 above)
Comment (10xLaCroixDrinker) [reply packages/holocron/__tests__/HolocronErrorBoundary.spec.jsx]: I dont see any assertions on the render as the spec suggests here either

File: packages/holocron/__tests__/moduleRegistry.spec.js [hunk 5]
```diff
@@ -38,7 +38,7 @@ describe('moduleRegistry', () => {
     const GoodModule = () => null;
     registerModule('good-module', GoodModule);
     expect(getModules()).toMatchSnapshot();
-    expect(getModule('good-module')).toBe(GoodModule);
+    expect(getModule('good-module').displayName).toBe('WithHolocronErrorBoundary(GoodModule)');
```
Comment (10xLaCroixDrinker) [inline packages/holocron/__tests__/moduleRegistry.spec.js]: any better option than asserting on the `displayName`?

File: packages/holocron/src/moduleRegistry.js [hunk 6]
```diff
@@ -31,7 +32,7 @@ function getModuleBlockList() {
 }
 
 function registerModule(moduleName, module) {
-  modules = modules.set(moduleName, module);
+  modules = modules.set(moduleName, withHolocronErrorBoundary(module));
```
Comment (10xLaCroixDrinker) [inline packages/holocron/src/moduleRegistry.js]: @PixnBits do you think users should be able to opt out of this?

File: packages/holocron/__tests__/HolocronErrorBoundary.spec.jsx [hunk 7]
```diff
@@ -0,0 +1,136 @@
+import React from 'react';
+// eslint-disable-next-line import/no-extraneous-dependencies
```
Comment (10xLaCroixDrinker) [inline packages/holocron/__tests__/HolocronErrorBoundary.spec.jsx]: why is this needed?

File: packages/holocron/src/HolocronErrorBoundary.jsx [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+  }
+
+  static getDerivedStateFromError(error) {
+    return { error };
+  }
+
+  componentDidCatch(error, errorInfo) {
+    this.setState({
+      error,
+    });
```
Comment (10xLaCroixDrinker) [inline packages/holocron/src/HolocronErrorBoundary.jsx]: this isn't needed with `getDerivedStateFromError`

File: packages/holocron/__tests__/HolocronErrorBoundary.spec.jsx (hunk 7 above)
Comment (JamesSingleton) [reply packages/holocron/__tests__/HolocronErrorBoundary.spec.jsx]: This is needed because `enzyme` is at the parent package.json level and not the `holocron` package.json level. It is done the same way here https://github.com/americanexpress/holocron/blob/master/packages/holocron/__tests__/RenderModule.spec.jsx#L17-L18

File: packages/holocron/src/moduleRegistry.js (hunk 6 above)
Comment (JamesSingleton) [reply packages/holocron/src/moduleRegistry.js]: Out of curiosity, why? This is a last line of defense from a module taking down a page if they don't implement their own error boundary. So if they opt out and don't implement their own, then a module would take down the page right?

File: packages/holocron/src/HolocronErrorBoundary.jsx (hunk 8 above)
Comment (JamesSingleton) [reply packages/holocron/src/HolocronErrorBoundary.jsx]: @10xLaCroixDrinker are you fine if I add `// eslint-disable-next-line class-methods-use-this` above the `componentDidCatch(error, errorInfo)` then?

File: packages/holocron/src/HolocronErrorBoundary.jsx (hunk 2 above)
Comment (JamesSingleton) [reply packages/holocron/src/HolocronErrorBoundary.jsx]: Taken care of in b6e3011

File: packages/holocron/__tests__/HolocronErrorBoundary.spec.jsx (hunk 3 above)
Comment (JamesSingleton) [reply packages/holocron/__tests__/HolocronErrorBoundary.spec.jsx]: Taken care of in b6e3011

File: packages/holocron/__tests__/HolocronErrorBoundary.spec.jsx (hunk 4 above)
Comment (JamesSingleton) [reply packages/holocron/__tests__/HolocronErrorBoundary.spec.jsx]: Taken care of in b6e3011

File: packages/holocron/src/HolocronErrorBoundary.jsx (hunk 8 above)
Comment (JamesSingleton) [reply packages/holocron/src/HolocronErrorBoundary.jsx]: Taken care of in b6e3011

File: packages/holocron/src/HolocronErrorBoundary.jsx (hunk 1 above)
Comment (JamesSingleton) [reply packages/holocron/src/HolocronErrorBoundary.jsx]: Taken care of in 42a0ed6

File: packages/holocron/src/moduleRegistry.js (hunk 6 above)
Comment (JAdshead) [reply packages/holocron/src/moduleRegistry.js]: what if they want to bubble up to error boundary in root module ?

File: packages/holocron/__tests__/HolocronErrorBoundary.spec.jsx [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+  const { NODE_ENV } = process.env;
+
+  beforeAll(() => {
+    consoleErrorSpy = jest.spyOn(console, 'error').mockImplementation(() => {});
+    mockError = new Error(
+      'These are not the droids you are looking for.'
+    );
+
+    // eslint-disable-next-line react/prop-types
+    Character = ({ name }) => {
```
Comment (JAdshead) [inline packages/holocron/__tests__/HolocronErrorBoundary.spec.jsx]: this can be moved out from the before all

Comment (JAdshead) [inline packages/holocron/__tests__/HolocronErrorBoundary.spec.jsx]: add prop `throwError` or something so that is explicit when it should throw or not

Comment (JamesSingleton) [reply packages/holocron/__tests__/HolocronErrorBoundary.spec.jsx]: Taken care of in 3649138

Comment (JamesSingleton) [reply packages/holocron/__tests__/HolocronErrorBoundary.spec.jsx]: Taken care of in 3649138

File: packages/holocron/__tests__/moduleRegistry.spec.js (hunk 5 above)
Comment (JamesSingleton) [reply packages/holocron/__tests__/moduleRegistry.spec.js]: Resolved in 291928d

File: packages/holocron/src/moduleRegistry.js (hunk 6 above)
Comment (JamesSingleton) [reply packages/holocron/src/moduleRegistry.js]: Or give their own fallback, or give their own `onError` function.

**Comment by anescobar1991** [comment]
No docs update?

**Comment by mtomcal** [comment]
Will there be an update to this to opt-out?

Also, what about adding to Holocron API docs with this new HOC?

File: packages/holocron/README.md [hunk 10]
```diff
@@ -14,6 +14,8 @@
 
 ## ✨ Features
 
+* Adds an ErrorBoundary around every Module to prevent a single Module from taking down a page once an error occurs in the module
+
```
Comment (mtomcal) [inline packages/holocron/README.md]: I know this effort was about error boundaries but it will look odd when merged to master showing Holocron having only one feature.

File: packages/holocron/src/moduleRegistry.js (hunk 6 above)
Comment (mtomcal) [reply packages/holocron/src/moduleRegistry.js]: - [ ] Need @PixnBits feedback

**Comment by PixnBits** [comment]
> This wraps modules being added to the module map with an ErrorBoundary HoC

Why here, instead of the `compose(.... errorBoundaryWrapper, ...)(Component)` in use in v4? A big downside of being here is the hidden cleverness/unknown magic instead of the predictable, explicitness of the v4 pattern.

File: packages/holocron/README.md (hunk 10 above)
Comment (JamesSingleton) [reply packages/holocron/README.md]: @mtomcal,

I agree. If you and the team could give some more features to add here I can. I believe the `ErrorBoundary` should be under Features and not under another part of the documentation.

**Comment by JamesSingleton** [comment]
I am going to decline this after further discussion with @PixnBits and @10xLaCroixDrinker due to the change being too opinionated.
